docs/developer/editors.md

```diff
@@ -0,0 +1,175 @@
+
+Editors all need to know which repository or repositories to work on. You can get additional parameters in the invocation as well, by creating your own Parameters class that extends BaseEditorOrReviewerParameters. The same ProjectEditor function works for both.
+
+See the [UpdateCopyright example](https://github.com/atomist/automation-client-samples-ts/blob/1da17e847b8e4a55ec246dfac351334ca49f3e71/src/commands/editor/addMyFile.ts) for an example of each.
+
+### Pull Requests or commits
+
+The `editMode` passed in the details to `editorHandler` determines how to save the changes. Pass an instance of [PullRequest](https://atomist.github.io/automation-client-ts/classes/_operations_edit_editmodes_.pullrequest.html), or an implementation of [BranchCommit](https://atomist.github.io/automation-client-ts/interfaces/_operations_edit_editmodes_.branchcommit.html). You can also supply a function from parameters to one of these EditModes, in case you want the branch name or commit message to vary by invocation.
+
+## Testing a project editor
+
+ProjectEditor is a function, so it's testable. Unit testing at this level is quick and powerful.
+
+### InMemoryProject
+
+Construct a project as input to your ProjectEditor. The simplest way is to list all the files and their content:
+
+```typescript
+import { InMemoryProject } from "@atomist/automation-client/project/mem/InMemoryProject";
+
+const startingProject = InMemoryProject.of({ path: "src/file1.ts", "All the stuff in file1"},
+                                           { path: "test/file2.ts", "All the stuff in file2"});
+```
+
+If your editor uses the `id` property on the Project, populate that too, using `InMemoryProject.from`:
+
+```typescript
+import { InMemoryProject } from "@atomist/automation-client/project/mem/InMemoryProject";
+
+const startingProject = InMemoryProject.from({ owner: "atomist", repo: "end-user-documentation", sha: "my-branch" },
+                                             { path: "src/file1.ts", "All the stuff in file1"},
+                                             { path: "test/file2.ts", "All the stuff in file2"});
+```
+
+Here is an [example](https://github.com/atomist/automation-client-samples-ts/blob/1da17e847b8e4a55ec246dfac351334ca49f3e71/test/commands/editor/addMyFileTest.ts) of a mocha test for the addMyFile project editor:
+
+```typescript
+it("adds the file", (done) => {
+    const p = InMemoryProject.of({ path: "existingFile.txt", content: "Hi\n" });
+    addMyFile(p).then(r => {
+            assert(p.findFileSync("myFile.txt")getContentSync().includes("I was here");
+        }).then(() => done(), done);
+});
+```
+
+## Working with Projects
+
+The [Project](https://atomist.github.io/automation-client-ts/interfaces/_project_project_.project.html) abstraction aims to provide both synchronous and asynchronous ways of changing projects. The synchronous methods simplify testing,
+while the asynchronous ones will give better performance in your automations.
```

Comment:
"the asynchronous ones will give better performance"
that seems like a bit of an oversimplification
"the asynchronous ones will give you more control over concurrency"
Maybe even reference that this is important due to many project operations being file or even network based and therefor I/O bound.